src/mbgl/style/sources/geojson_source_impl.cpp

```diff
@@ -18,45 +19,39 @@ class GeoJSONVTData : public GeoJSONData {
         : impl(geoJSON, options) {
     }
 
-    mapbox::feature::feature_collection<int16_t> getTile(const CanonicalTileID& tileID) final {
-        return impl.getTile(tileID.z, tileID.x, tileID.y).features;
+    void getTile(const CanonicalTileID& id, std::function<void(TileFeatures)> fn) final {
+        assert(fn);
+        // It's safe to pass `this` as scheduler will die earlier.
+        scheduler.scheduleAndReplyValue(
+            [id, this]() -> TileFeatures { return impl.getTile(id.z, id.x, id.y).features; }, std::move(fn));
     }
 
-    mapbox::feature::feature_collection<double> getChildren(const std::uint32_t) final {
-        return {};
-    }
+    Features getChildren(const std::uint32_t) final { return {}; }
 
-    mapbox::feature::feature_collection<double>
-    getLeaves(const std::uint32_t, const std::uint32_t, const std::uint32_t) final {
-        return {};
-    }
+    Features getLeaves(const std::uint32_t, const std::uint32_t, const std::uint32_t) final { return {}; }
 
     std::uint8_t getClusterExpansionZoom(std::uint32_t) final {
         return 0;
     }
 
 private:
     mapbox::geojsonvt::GeoJSONVT impl;
+    SequencedScheduler scheduler;
```

Comment:
why not to use dedicated thread pool? This will create new thread for each geojson source, right?

Reply:
we need a single thread that is keeping the state. Right, for each geojson source impl.